core/src/agent/state.rs

```diff
@@ -30,6 +39,17 @@ pub struct AgentState {
     top_chain_header: Option<ChainHeader>,
 }
 
+#[derive(Debug, Deserialize, Serialize)]
+pub struct AgentStateSnapshot {
+    top_chain_header: ChainHeader,
+}
+
+impl AgentStateSnapshot {
+    pub fn top_chain_header(self) -> ChainHeader {
```

Comment:
This method can only be called once 😁 
```suggestion
    pub fn top_chain_header(&self) -> ChainHeader {
```

Reply:
+1